### ISSUE TITLE
`Type.as_obj` crashes on bases

### ISSUE THREAD

**Issue by ggreif** [dfinity/motoko#3405]
Below problem surfaced with 23b4e8bca:
``` Motoko
func mux<A <: { a : Int }, B <: { b : Char }>(a : A, b : B) : { a : Int; b : Char; c : Text } =
    { a and b with c = "Yeah" }
```
I get a crash:
```
+OOPS! You've triggered a compiler bug.
+Please report this at https://github.com/dfinity/motoko/issues/new with the following details:
+
+Motoko (source XXX)
+
+Fatal error: exception (Invalid_argument Type.as_obj)
```

**Comment by ggreif** [comment]
The problem is that `Type.as_obj` cannot handle bases when bounded.
``` Motoko
func mux<A <: { a : Int }, B <: { b : Char }>(a : A, b : B) : { a : Int; b : Char; c : Text } =
    { a with b = 'B'; c = "Yeah" } //     { a = 8; b = 'B'; c = "Yeah" } <- works
```

**Comment by ggreif** [comment]
Fixed by #3417.
